test/integration/replicaset/replicaset_test.go

```diff
@@ -268,250 +510,438 @@ func TestAdoption(t *testing.T) {
-	}); err != nil {
-		t.Fatal(err)
+	// Polling is not needed as 5 seconds already passed
+	if newRS.Status.AvailableReplicas != 1 {
+		t.Fatalf("Unexpected .Status.AvailableReplicas: Expected 1, saw %d", newRS.Status.AvailableReplicas)
 	}
-	close(stopCh)
 }
 
-// selectors are IMMUTABLE for all API versions except extensions/v1beta1
-func TestRSSelectorImmutability(t *testing.T) {
-	s, closeFn, clientSet := rmSimpleSetup(t)
+func TestRSScaleSubresource(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
 	defer closeFn()
-	ns := framework.CreateTestingNamespace("rs-selector-immutability", s, t)
+	ns := framework.CreateTestingNamespace("test-rs-scale-subresource", s, t)
 	defer framework.DeleteTestingNamespace(ns, s, t)
-	rs := newRS("rs", ns.Name, 0)
-	createRSsPods(t, clientSet, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+	defer close(runControllerAndInformers(rm, informers))
 
-	// test to ensure extensions/v1beta1 selector is mutable
-	newSelectorLabels := map[string]string{"changed_name_extensions_v1beta1": "changed_test_extensions_v1beta1"}
-	rs.Spec.Selector.MatchLabels = newSelectorLabels
-	rs.Spec.Template.Labels = newSelectorLabels
-	replicaset, err := clientSet.ExtensionsV1beta1().ReplicaSets(ns.Name).Update(rs)
+	rs := newRS("rs", ns.Name, 1)
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{})
+	rs = rss[0]
+	waitRSStable(t, c, rs)
+
+	// Use scale subresource to scale up .Spec.Replicas to 3
+	testScalingUsingScaleSubresource(t, c, rs, 3)
+	// Use the scale subresource to scale down .Spec.Replicas to 0
+	testScalingUsingScaleSubresource(t, c, rs, 0)
+}
+
+func TestExtraPodsAdoptionAndDeletion(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-extra-pods-adoption-and-deletion", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	defer close(runControllerAndInformers(rm, informers))
+
+	rs := newRS("rs", ns.Name, 2)
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{})
+	rs = rss[0]
+	waitRSStable(t, c, rs)
+
+	// Create another pod that the RS wants to adopt
+	pod := newMatchingPod("extra-pod", ns.Name)
+	pod.Labels = labelMap()
+	podClient := c.Core().Pods(ns.Name)
+	_, err := podClient.Create(pod)
 	if err != nil {
-		t.Fatalf("failed to update extensions/v1beta1 replicaset %s: %v", replicaset.Name, err)
-	}
-	if !reflect.DeepEqual(replicaset.Spec.Selector.MatchLabels, newSelectorLabels) {
-		t.Errorf("selector should be changed for extensions/v1beta1, expected: %v, got: %v", newSelectorLabels, replicaset.Spec.Selector.MatchLabels)
+		t.Fatalf("Failed to create pod: %v", err)
 	}
 
-	// test to ensure apps/v1beta2 selector is immutable
-	rsV1beta2, err := clientSet.AppsV1beta2().ReplicaSets(ns.Name).Get(replicaset.Name, metav1.GetOptions{})
-	if err != nil {
-		t.Fatalf("failed to get apps/v1beta2 replicaset %s: %v", replicaset.Name, err)
+	// Verify there are 3 pods
```

Comment:
This is flaky. The RS might beat you and delete the Pod before you get here. This might be a good time to pre-create the Pods; then you will know there were 3 Pods before starting the RS controller, and you can Poll until there are 2. Make sure you fix the ordering of the call to `waitToObservePods` as mentioned above.

Reply:
Ya, there is a race condition. +1 on pre-creating the pods.